src/backend/cuda/jit.cpp

```diff
@@ -233,82 +260,138 @@ void evalNodes(vector<Param<T>> &outputs, const vector<Node *> &output_nodes) {
     }
 
     bool is_linear = true;
-    for (auto node : full_nodes) { is_linear &= node->isLinear(outDims); }
+    for (const auto &node : full_nodes) {
+        is_linear &= node->isLinear(outDims);
+    }
+    for (int dim = 0, elements = 1; dim < ndims; ++dim) {
+        is_linear &= (elements == (int)outputs[0].strides[dim]);
+        elements *= (int)outDims[dim];
+    }
```

Comment:
yes, output's are currently linear for all the cases I recall.

Reply:
output is linear in case of eval, because a new buffer is allocated and provided to this function.

When reviewing this function, I discovered that all nodes have 2 modes of operation:
1. linear input, linear output
2. non-linear input, non-linear output (also used for linear output)

This brought me to the idea, that I also could write into a non-linear buffer or even a part of a buffer as long as provided the correct buffer parameters to the eval function.
This would replace
-  the write to the linear buffer (original eval) followed by a read from the linear buffer, followed by a write to the final buffer by 
- a direct write into the final buffer.

When you look at the join & copy method with non-evaluated inputs, you will see that the eval function will now write directly in place in the final buffer (non-linear buffer).  In jit, I only had to force the non-linear mode for non-linear result buffers.  As expected, this almost doubled the performance for these cases.

Conclusion: the functionality (writing in a non-linear buffer) already existed, although was never used.

Reply:
The loop is indeed, detecting if a buffer is non-linear.

Reply:
Yeah, join(0, arr, arr) will also create non-linear output array. This should help with that as well. :+1: